### ISSUE TITLE
Assembly search path should be more configurable to support reproducible builds

### ISSUE THREAD

**Issue by simonthum** [dotnet/msbuild#772]
Hi MSBuild team,

as briefly mentioned in #749 I stumbled over MSBuild's default assembly search path and lost quite some time with that. I was diagnosing an issue and deleted some internally provisioned assemblies to make sure I was getting the right ones, but wasn't due to MSBuild resolving dependencies from the OutputPath/OutDir. This was highly surprising to me, and in fact it's problematic in our build since we have other tools in play.

What I would expect is that MSBuild is transforming a set of inputs to a set of outputs. Resolving assemblies from the OutputPath is conflating the concepts and is creating problems in workflows where the output path is regularly populated by other tools. I do have a fix, but it's ugly.

The larger issue is that of reproducible builds. As you're aware there are a number of locations in the AssemblySearchPath. If I were to make an architectural decision to depend only on Framework and internally provisioned assemblies (i.e. no arbitrary GAC contents etc.), I would need to go out of my way to check and document that (In fact we have the moral equivalent of this decision due to our target platform) and still any dev could break it in a rush without even noticing.

If I could limit the search path in msbuild, or have the assembly resolution task fail or warn on project-designated "bad" sources we'd be in a better position. The bad sources could be greyed in VS to make it abundantly clear to devs what's going on.

**Comment by rainersigwald** [comment]
@nguerrera This idea seems like it'd play well with some of the things you're working on, especially around cross-targeting from desktop MSBuild (but still making sure that references come from packages and not other locations).

**Comment by nguerrera** [comment]
You *can* override AssemblySearchPaths so I'm not sure what the request is here.

I've made the default for our new project system exclude GAC, registry-based locations, and output directories. I sympathize with the spirit here because all of those have caused me nothing but pain over the years. :)

**Comment by simonthum** [comment]
@nguerrera In summary, the request is to make it easier to control assembly resolution or change AssemblySeachPath.

Yes I can do that now, but if I want to retain the ability to resolve against a target framework (i.e. externally provisioned metadata-only assemblies) then I'd have to know where they are to re-add them somehow or remove the paths I don't want by matching them character-by-character in my build files.

I'm doing the latter, but one day the characters introducing that path are going to change and I know I won't note.

If you know a better way to achieve the above, please let me know!

**Comment by cdmihai** [comment]
I'll add toggles for each line in https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L565-L576.
That way people can have fine grained control, and the SDK can use the toggles instead of rewriting the search paths.

**Comment by rainersigwald** [comment]
I think it'd be nice to have a single toggle for what we think is right, too. That'd make it easy to make an existing project more robust in this way, and not require converting to the SDK or setting several toggles.

**Comment by cdmihai** [comment]
Is this what we think is right? `{HintPathFromItem};{RawFileName}`
More lenient: `{HintPathFromItem};{TargetFrameworkDirectory};{RawFileName}`
Does `TargetFrameworkDirectory` even make sense for .net core? Should we have two good ones, one for full framework and one for .net core?

For reference, the SDK has these: https://github.com/dotnet/sdk/blob/6945e3694c918eea4c8c4fb6217e1485b179994b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props#L93

And I guess we won't touch the design time search paths, as that's VS' property: https://github.com/Microsoft/msbuild/blob/master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L2621. They're more restrictive than the non design time ones.

So, I guess my proposal is now:
- add a toggle for using a "machine independent" set. Options: `UseRestrictedRARSearchPaths` / `UseMinimalRARSearchPaths` / `UseExplicitRARSearchPaths` (meaning there's no search, something points directly to the path on disk)
- also add a toggle for each line in the current ones. I don't know how useful this is. Sounds useful only if we get the minimal set wrong, and users want to tweak.

**Comment by rainersigwald** [comment]
I think we need to keep `{TargetFrameworkDirectory}` to support bare references to framework types, like

```xml
<Reference Include="System.Xml" />
```

Those are extremely common in .NET Framework-targeting projects, and it should be safe enough since we ensure the targeting pack for the target framework is installed.

**Comment by nguerrera** [comment]
We haven't had any problems with TargetFrameworkDirectory. We always have it set in Core, but we clear out the directories: https://github.com/dotnet/sdk/blob/7aaaeaf06d49047648609337b0ecd1a2b8cf97b8/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DisableStandardFrameworkResolution.targets#L17-L26

**Comment by rainersigwald** [comment]
I'm with Nick on `{HintPathFromItem}`: put it first; leave it in. Too much Internet Code that uses it, because VS has long used it, even if we change VS to not  use it any more.

**Comment by cdmihai** [comment]
We decided to just add better documentation on how to set assembly search paths as opposed to introducing new toggles, since the toggles would be more complicated than just setting the right property to the right values.

**Comment by simonthum** [comment]
@cdmihai That sounds reasonable.

But please provide examples, I've always found it painful to work with ASP, e.g. when I had to fight an SDK that did not like to be installed in applicable project folders by failing to resolve native assemblies.

**Comment by aaronla-ms** [comment]
Maybe we could start a wiki or doc to record all the settings that users should configure on dotnetsdk in order to set themselves up for reproducible builds?

**Comment by aaronla-ms** [comment]
@nguerrera 

> I've made the default for our new project system exclude GAC, registry-based locations, and output directories. I sympathize with the spirit here because all of those have caused me nothing but pain over the years. :)

Nick, can you point me to this new project system? We just moved to sdk style `<Project Sdk="Microsoft.NET.Sdk">`, but are still seeing GAC on AssemblySearchPath by default when targetting netframework. 

> You can override AssemblySearchPaths so I'm not sure what the request is here

True, and I'm definitely thankful that we can override this setting. But it's hard to determine the full set of settings needed to configure a reproducible build for even the simplest projects. Maybe a doc on reproduceable builds that documents all the settings needed to achieve this? Or perhaps a nuget package that checks for anti-reproducible settings and warns for them, so users can know if they've checked a good set of options? 

I'm just spit-balling some ideas here.

**Comment by jaredpar** [comment]
> The larger issue is that of reproducible builds. As you're aware there are a number of locations in the AssemblySearchPath. If I were to make an architectural decision to depend only on Framework and internally provisioned assemblies (i.e. no arbitrary GAC contents etc.), I would need to go out of my way to check and document that (In fact we have the moral equivalent of this decision due to our target platform) and still any dev could break it in a rush without even noticing.

This problem isn't one I would prescribe to build reproducibility. 

Generally *reproducible* builds are builds that [meet the qualification](https://reproducible-builds.org/) of given the same inputs they will produce the same outputs (byte for byte). The fact that `$(AssemblySearchPath)` looks in a lot of places on the machine for resolution is immaterial to reproducibility. If the environment is the same, that is the machine state has the same set of files in all of the places considered by `$(AssemblySearchPath)`, then MSBuild will pick the same assemblies (same input, same output). 

The request here is more around build *isolation*. Essentially the fact that `$(AssemblySearchPath)` looks for artifacts on the machine that weren't either a part of the .NET SDK or produced by my build means unexpected machine state is factoring into my build (my build is not isolated from other machine state). Removing say the GAC from `$(AssemblySearchPaths)` isn't changing the *reproducibility* of your build, it's just changing the state which is part of the "input" to your build. 

This may seem like a pedantic point to make but it's an important one. The .NET SDK meets the *reproducibility* requirement for builds in the core scenarios. That is **very** important to us. The ability to *isolate* builds from other state on the machine is in many cases supported but not as well documented and understood. For example I couldn't rattle off all the extension points of MSBuild :smile: 

I'm not trying to diminish the bug report here or the arguments. I would very much like it if were easier to configure MSBuild to be more isolated. But that's separate from whether or not it's outputs are reproducible. 

> I would love there to be an opt-in mode where msbuild worked soley off the installed SDK and nugets, and ignored everything else on the system.

This is essentially an ask for a very strict *isolated* mode for MSBuild.  Again not disagreeing, but it's different than *reprocubility*. 

> Maybe we could start a wiki or doc to record all the settings that users should configure on dotnetsdk in order to set themselves up for reproducible builds?

I think that would be very valuable. You listed a good number of the ones above but there are more. For example you need to ensure that you use a different NuGet folder if you want to be truly isolated from other builds on the machine. Having a shared place where we could list off all the items that impact isolation, and the pros / cons of flipping them, would likely be beneficial to a number of build maintainers in .NET.

**Comment by aaronla-ms** [comment]
> If the environment is the same, that is the machine state has the same set of files in all of the places considered by `$(AssemblySearchPath)`, then MSBuild will pick the same assemblies (same input, same output).

Yes, but one of the principles of reproducible-builds.org for build systems is "Capture as little as possible from the environment". While you _can_ say that your minimal environment is the entirety of everything installed on your build machine, it's very much against the intent here. Why depend on the GAC at all when there are fine alternatives available in the form of self contained reference assembly packages that can be restored deterministically?

**Comment by nguerrera** [comment]
> Nick, can you point me to this new project system? We just moved to sdk style <Project Sdk="Microsoft.NET.Sdk">, but are still seeing GAC on AssemblySearchPath by default when targetting netframework.

@aaronla-ms Hmm, I suspect there's something else in your projects causing this. I just checked with the following csproj:

``` xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
  </PropertyGroup>
</Project>
```

I get the following AssemblySearchPaths:

![image](https://user-images.githubusercontent.com/75470/115476463-c1c6f380-a1f6-11eb-9067-51a0dcd0a4b2.png)

I also don't see anything conditioning this on TargetFramework:

https://github.com/dotnet/sdk/blob/4b47852bda23ec4fed8bcadf8186aa5cfb217cd7/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props#L91

If you can share repro steps or a https://aka.ms/binlog, we can probably chase down what's happening.

**Comment by aaronla-ms** [comment]
Ah, you're right; in our latest master, that's gone. 

We do see TargetFrameworkDirectory gets auto-resolved to sniff out registry and program files still, but it looks like i can remove <AssemblySearchPath> from our IsolatedBuild.props, and just keep the override of `TargetFrameworkPath=[undefined]`

**Comment by aaronla-ms** [comment]
@jaredpar 

> Removing say the GAC from $(AssemblySearchPaths) isn't changing the reproducibility of your build, it's just changing the state which is part of the "input" to your build.

> This may seem like a pedantic point to make but it's an important one. The .NET SDK meets the reproducibility requirement for builds in the core scenarios.

Perhaps you've had different experience, but I've it to be unrealistic to assume users will have the same Visual Studio or GAC configuration. E.g. some users are running VS 2019 Preview, some are using VS 2017 because they don't really use VS much, and yet others have some weird versions of Newtonsoft.Json installed in their GAC because some installer thought that was a Good Idea.

While it could work in theory, I just haven't found it to last long in practice. We've had better success by pruning dependencies on Program Files, GAC, and registry where practical.

**Comment by jaredpar** [comment]
> Yes, but one of the principles of reproducible-builds.org for build systems is "Capture as little as possible from the environment". While you can say that your minimal environment is the entirety of everything installed on your build machine, it's very much against the intent here.

Completely agree. Ideally you should focus on adding as much isolation to the build as possible. All the production builds I maintain attempt to do this. At the end of the day though there is a difference between problems of isolation and reproducibility. That is the main point I wanted to get across because they are diff problems and have diff priorities, requirements, etc ... attached to them. 

> Why depend on the GAC at all when there are fine alternatives available in the form of self contained reference assembly packages that can be restored deterministically?

There are historic reasons why MSBuild needs to do this. It's similar to asking "why is the default of the C# compiler to have non-deterministic builds?" The answer, for both cases, is because if we did the "better" thing it would break customers when they moved to a new version of Visual Studio. The only realistic option is to provide knobs, and ideally permanent opt-ins, to let customers who care get on the golden path for these types of scenarios.

**Comment by simonthum** [comment]
@jaredpar The source you linked say "Reproducible builds are a set of software development practices that create an _independently-verifiable_ path from source to binary code." (emphasis mine)

As is explained in more detail under "How?", the concept of isolation is part of that definition, they also advocate for a way to recreate a "close enough" environment.

This leaves us with the two aspects you mention:
* deterministic, idempotent or reproducible in the narrow sense: AFICT in roslyn this is realized.
* ability to recreate, pre-define or isolate the build: This still needs better msbuild support.

For clarity, let's stick with the notion that reproducible builds need at least both.

@aaronla-ms I'm happy to share my snippets, but it seems you already found more tricks than me. BTW I dropped OutDir and never missed it since.

**Comment by jaredpar** [comment]
> ability to recreate, pre-define or isolate the build: This still needs better msbuild support.

Only the last item there needs better build support. Technologies like docker or simple setup instructions can solve the first two without changes to msbuild.

**Comment by aaronla-ms** [comment]
> @aaronla-ms I'm happy to share my snippets, but it seems you already found more tricks than me. BTW I dropped OutDir and never missed it since.

You give  me too much credit! But yeah, I feel your pain too. 

Please feel free to share what you've found so far. Open communication is the best way we can mutually learn in this sort of ad-hoc design space :)

**Comment by simonthum** [comment]
@jaredpar  I agree, I binned those together solely by purpose.

I think a collection of build isolation settings and tricks would get the job done, most of the time. I'll post what I manage to find in my project files.

**Comment by jaredpar** [comment]
> I think a collection of build isolation settings and tricks would get the job done, most of the time. I'll post what I manage to find in my project files.

Would really like to see this. If we can get a doc going somewhere I'm more than happy to help contrib. Know we've accumulated a few settings over the years.

I'm also interested to see how we eventually categorize settings. I'm guessing items like removing GAC from the `$(AssemblySearchPaths)` is likely to be fairly non-controversial. Removing that is very important to maintaining build isolation. But what about the NuGet restore directory? If you want true isolation then that needs to be solution specific otherwise you can get into cross feed poisoning issues (unless you use lock files). But I'm guessing some teams would probably end up sharing this directory because of the perf gains associated with doing so.

**Comment by aaronla-ms** [comment]
> If you want true isolation then that needs to be solution specific otherwise you can get into cross feed poisoning issues (unless you use lock files). But I'm guessing some teams would probably end up sharing this directory because of the perf gains associated with doing so.

Oh, it gets worse than that. Our old build infrastructure doesn't permit per-repo package directories, and at time has restored the _wrong package contents_ when some peer team had a private feed with a manually modified version of the nuget under the same name and version. And forget about cross repo poisoning -- I've seen dozens of peers fail to notice feed configuration errors because a previous commit of that repo had restored packages into the cache.

I'm currently recommending my peers use a shared cache, but merge validation builds _always_ use a clean cache. That way devs don't have to pay the restore cost, but we don't regress main.

**Comment by Forgind** [comment]
I think resolved via #7008? I didn't read all the conversation, so let me know if not.
